### PR TITLE
[PyTorch] Fix mutable QB bounds in CUDA graphs

### PR DESCRIPTION
**Pull request by harryzhou2000**
## Rebase update (2026-09-07)

Rebased onto current `upstream/main` at `5f6105b9` (`[Common] Preserve shared-memory pointer provenance in TMA kernels (#3482)`). The five-commit series rebased cleanly; the new head is `30d9abfa`.

The implementation is not Hopper- or Blackwell-specific: the final diff remains Python and tests only, has no architecture guards or architecture-specific kernel changes, and uses standard CUDA graph/tensor semantics. After the rebase it was built and validated in `nvcr.io/nvidia/pytorch:26.04-py3` on 2x NVIDIA A100 80GB PCIe (SM80):

- Editable PyTorch TE build completed successfully for `NVTE_CUDA_ARCHS=80`.
- Focused mutable-bounds CUDA-graph tests: `12 passed`.
- Complete fused-router QB selection: `47 passed, 2 skipped`; both histogram modes passed. The two skipped tests require multiple visible GPUs, while this pytest invocation intentionally exposed one GPU.
- Final focused post-environment-check rerun: `4 passed`.
- Both A100s were visible in the container; dependency validation reported no broken requirements.

## Description

Quantile Balancing keeps `qb_bin_bounds` as a persistent CUDA tensor. In MCore's full-iteration CUDA graph, the router reads the current bounds near the start of an iteration and QB finalization updates the same storage near the end of that iteration. The graph executable is stable, but the state behind its captured device pointer is intentionally mutated once per replayed global batch.

The host validation cache introduced in https://github.com/NVIDIA/TransformerEngine/pull/3395 keys eager validation to the tensor's PyTorch version. Each eager `bounds.copy_()` advances that version while preserving the pointer, so the last warmup leaves a valid but stale cached version and capture attempts a forbidden device-to-host validation.

Graph replay has a second important property:

- dispatching the in-place update while capturing advances `_version` once;
- each later `cudaGraphLaunch` executes the captured update and changes device contents;
- replay does not re-enter Python, so `_version` remains fixed.

The fix therefore uses an explicit trusted-producer handshake rather than pretending host version bookkeeping can inspect replayed data:

- eager router calls retain recoverable host validation and exact-version caching;
- `mark_qb_bin_bounds_validated()` records the current version without synchronizing or inspecting values;
- the marker is allowed immediately after a trusted in-place update is dispatched during capture on the same stream;
- capture still rejects a never-validated tensor or any stale version that was not explicitly produced by the trusted updater;
- replay reads later values through the stable pointer under the updater's validity contract.

MCore's QB updater is such a trusted producer: starting from finite ordered bounds, it computes finite expert biases and constructs the next range as `[bias_min - 1, bias_max + 1]`. The companion feature-detected MCore call is https://github.com/harryzhou2000/Megatron-LM/commit/78e971efa.

This PR intentionally does not use a device assertion. Invalid replay-time data cannot become a recoverable Python exception without a caller-owned device status tensor and a later host observation point; an assertion or trap would poison the CUDA context. If an untrusted graph-time producer is added later, it should use that explicit status-buffer design. The current trusted updater needs no extra kernel, global write, or replay-time host synchronization.

The final Transformer Engine diff is Python and tests only; it does not change a CUDA kernel, C++ binding, public C ABI, or fused-router launch sequence. Kimi K3's Quantile Balancing algorithm is described in https://github.com/MoonshotAI/Kimi-K3/blob/main/k3_tech_report.pdf.

## Tests

Previously validated on one NVIDIA B200 using the `dev_2604` container:

- Focused QB bounds/CUDA-graph selection: `22 passed`.
- Complete fused-router QB selection: `47 passed, 2 skipped`; the skips require multiple GPUs.
- Both `two_kernel` and `fused_atomic` histogram modes.
- Sparse routing-map and caller-provided dense Top-k output paths.
- Exact-version rejection for never-validated and stale pre-capture bounds.
- Capture with a trusted bounds update after router execution, matching full-iteration ordering.
- Production-shaped MCore integration captured routing and QB finalization in one graph, then replayed four times. The bounds pointer stayed stable; bounds changed on every replay and remained finite/ordered; `_version` advanced once during capture and remained fixed across all replays.
- Targeted pre-commit hooks pass for all changed files.

### CODE REVIEW

File: transformer_engine/pytorch/router.py [hunk 1]
```diff
@@ -55,10 +55,15 @@ def _validate_qb_bin_bounds(bin_bounds: torch.Tensor) -> bool:
...
-    if getattr(bin_bounds, _QB_BOUNDS_VALIDATED_VERSION_ATTR, None) == version:
+    validated_version = getattr(bin_bounds, _QB_BOUNDS_VALIDATED_VERSION_ATTR, None)
+    if validated_version == version:
         return True
     with torch.cuda.device(bin_bounds.device):
         if torch.cuda.is_current_stream_capturing():
+            if validated_version is not None:
+                # CUDA graphs support persistent input tensors whose values are updated in place
+                # between capture and replay. The kernel reads the current bounds by device pointer.
+                return True
```
[the remark this answers is not in the document]
Comment (harryzhou2000) [reply transformer_engine/pytorch/router.py]: Thanks. This is a real distinction for arbitrary callers, but it is also inherent to mutable CUDA-graph inputs: after capture, replay never re-enters this Python helper, so an invalid in-place update between any two replays would bypass host validation in exactly the same way. The production MCore contract updates this persistent tensor once outside the captured forward/backward using the QB updater, which preserves finite ordered bounds.

A device assertion is not a recoverable substitute: Greptile correctly rejected that earlier design in https://github.com/NVIDIA/TransformerEngine/pull/3395#discussion_r3802301171 because it poisons the CUDA context. I also verified `torch._assert_async` has the same behavior during both capture and replay. A recoverable per-replay validation would require a new device status output and an explicit host check after graph replay, which is a cross-repository API/protocol change rather than this minimal regression fix.

Commit 2c69379c now documents the contract at both the capture decision and the public argument: eager calls revalidate every PyTorch-tracked mutation, while callers of a captured graph must preserve valid values across capture and replay. The negative eager-update coverage remains in both histogram modes, and never-eager-validated tensors are still rejected during capture.

[the remark this answers is not in the document]
Comment (harryzhou2000) [reply transformer_engine/pytorch/router.py]: Follow-up: I implemented the stricter handshake in 8841e048 instead of relying only on documentation. Capture again requires the tensor's exact current version to be validated. A new public `mark_qb_bin_bounds_validated` lets a trusted device-side updater record that exact version outside capture without a host synchronization; it is explicit that the updater guarantees finite ordered values. The companion MCore update calls it immediately after the QB update whose construction guarantees `lower = bias_min - 1` and `upper = bias_max + 1`: https://github.com/harryzhou2000/Megatron-LM/commit/78e971efa.

The new negative test eagerly validates valid bounds, mutates them to invalid bounds, and verifies capture rejects the stale version in both histogram modes. Final validation is 20 focused TE graph tests, 45 full TE QB tests, 18 MCore QB unit tests, and the MCore three-eager-update plus capture plus three-replay integration.

File: transformer_engine/common/fused_router/fused_topk_with_score_function.cu [hunk 2]
```diff
@@ -36,6 +36,16 @@ struct QBBinParams {
 __device__ inline QBBinParams load_qb_bin_params(const CompType *bin_bounds, int num_bins) {
   const CompType lower = bin_bounds[0];
   const CompType upper = bin_bounds[1];
+  const bool valid = isfinite(lower) && isfinite(upper) && lower < upper;
+  if (!valid) {
+    if (threadIdx.x == 0 && threadIdx.y == 0 && threadIdx.z == 0) {
+      printf("QB bin_bounds values must be finite with lower < upper.\n");
+      __trap();
+    }
```
[the remark this answers is not in the document]
Comment (harryzhou2000) [reply transformer_engine/common/fused_router/fused_topk_with_score_function.cu]: Fixed on the current head (`583702dd`). The device validation kernel and `__trap()` path from `5b2d1e12` were removed, so the final PR has no device assertion, CUDA/C++ change, or ABI change. Capture now requires the tensor current version to be validated, with an explicit marker only for the trusted same-stream MCore QB update; invalid eager bounds and stale unmarked versions still fail recoverably before launch.

A genuinely untrusted graph-time producer cannot raise a recoverable Python exception from inside a CUDA kernel. Supporting that case would require a separate asynchronous status protocol: a caller-owned persistent device status tensor, device validation that records an error and skips QB histogram writes while preserving routing outputs, MCore finalization gated to retain the previous bias/bounds, and a collective host check at a safe post-graph synchronization point. That cross-repository API is intentionally outside this minimal regression fix. The final implementation therefore does not use a trap.

File: tests/pytorch/test_fused_router.py [hunk 3]
```diff
@@ -715,19 +723,152 @@ def run_router():
...
             qb_histogram_mode=histogram_mode,
         )
 
     run_router()
+    bounds_data_ptr = bin_bounds.data_ptr()
+    validated_version = bin_bounds._version
+    bin_bounds.copy_(torch.tensor([-0.25, 0.75], device="cuda"))
+    assert bin_bounds.data_ptr() == bounds_data_ptr
+    assert bin_bounds._version != validated_version
+    mark_qb_bin_bounds_validated(bin_bounds)
```
Comment (ksivaman) [inline tests/pytorch/test_fused_router.py]: Is mutating the bounds here actually a useful test? The use case should be to change the bounds after the capture and then ensure replay produces an accurate histogran

Comment (harryzhou2000) [reply tests/pytorch/test_fused_router.py]: Yes, agreed. The useful contract here is changing the bounds after capture and verifying that replay reads the updated values through the stable device pointer.

Fixed in `b7545d7c`: I removed the redundant pre-capture mutation/marker from this test. It now eagerly validates the initial bounds, captures the router, updates the same bounds tensor after capture, replays, and compares the histogram with the updated-bounds reference. This covers both histogram modes and dense/sparse routing output. The separate `test_qb_topk_cuda_graph_captures_bounds_update` continues to cover an updater captured inside the iteration.

While tracing this for general TE callers, I also made the replay boundary explicit in the Python docstrings: replay does not execute Python validation or advance PyTorch tensor-version counters, so replay-visible producers must preserve finite, ordered bounds for every input/control-flow path. Capture-time marker trust is now distinguished from ordinary eager validation, forcing the next eager router call to inspect replay-produced values again; a regression test verifies that an invalid captured update is rejected on that eager call.

Local validation on A100 SM80 after rebuilding the reviewer-updated branch: 12 focused CUDA-graph cases and all 51 QB tests passed; pinned Black and `pip check` also pass.

**Review by ksivaman — APPROVED** [review]
LGTM
